lib/knapsack_pro/formatters/time_tracker.rb

```diff
@@ -0,0 +1,162 @@
+# frozen_string_literal: true
+
+module KnapsackPro
+  module Formatters
+    class TimeTracker
+      ::RSpec::Core::Formatters.register self,
+        :example_group_started,
+        :example_started,
+        :example_finished,
+        :example_group_finished,
+        :stop
+
+      # Called at the beginning of each subset,
+      # but only the first instance of this class is used,
+      # so don't rely on the initializer to reset values.
+      def initialize(_output)
+        @time_each = nil
+        @time_all = nil
+        @before_all = 0.0
+        @group = {}
+        @subset = {}
+        @queue = {}
+        @suite_started = now
+        @subset_started = now
+      end
+
+      def example_group_started(notification)
+        return unless top_level_group?(notification.group)
+        @time_all = now
+      end
+
+      def example_started(_notification)
+        @before_all = now - @time_all if @before_all == 0.0
+        @time_each = now
+      end
+
+      def example_finished(notification)
+        record_example(@group, notification.example, @time_each)
+        @time_all = now
+      end
+
+      def example_group_finished(notification)
+        return unless top_level_group?(notification.group)
+
+        add_hooks_time(@group, @before_all, now - @time_all)
+        @subset = merge(@subset, @group)
+        @before_all = 0.0
+        @group = {}
+      end
+
+      # Called at the end of each subset
+      def stop(_notification)
+        @queue = merge(@queue, @subset)
+        @subset = {}
+        @subset_started = now
+      end
+
+      def queue(scheduled_paths)
+        recorded_paths = @queue.values.map do |example|
+          example[:path].match(/\A(.*?)(?:\[([\d\s:,]+)\])?\z/).captures.first
```

Comment:
We use this pattern `match(/\A(.*?)(?:\[([\d\s:,]+)\])?\z/).captures.first` in 2 places. Should we extract this somewhere? To the `get_spec_id` method or something like that?